lib/src/wallet/traits.rs

```diff
@@ -0,0 +1,555 @@
+use super::{
+    data::{OrchardNoteData, SaplingNoteData, WalletNullifier, WalletTx, WitnessCache},
+    keys::{orchard::OrchardKey, sapling::SaplingKey, Keys},
+    transactions::WalletTxns,
+};
+use nonempty::NonEmpty;
+use orchard::{
+    bundle::{Authorized as OrchardAuthorized, Bundle as OrchardBundle},
+    keys::{
+        Diversifier as OrchardDiversifier, FullViewingKey as OrchardFullViewingKey,
+        IncomingViewingKey as OrchardIncomingViewingKey,
+        OutgoingViewingKey as OrchardOutgoingViewingKey, SpendingKey as OrchardSpendingKey,
+    },
+    note::{Note as OrchardNote, Nullifier as OrchardNullifier},
+    note_encryption::OrchardDomain,
+    primitives::redpallas::{Signature, SpendAuth},
+    tree::MerkleHashOrchard,
+    Action, Address as OrchardAddress,
+};
+use zcash_address::unified::{self, Encoding as _, Receiver};
+use zcash_client_backend::encoding::encode_payment_address;
+use zcash_note_encryption::{Domain, ShieldedOutput, ENC_CIPHERTEXT_SIZE};
+use zcash_primitives::{
+    consensus::{BlockHeight, Parameters},
+    keys::OutgoingViewingKey as SaplingOutgoingViewingKey,
+    memo::{Memo, MemoBytes},
+    merkle_tree::Hashable,
+    sapling::{
+        note_encryption::SaplingDomain, Diversifier as SaplingDiversifier, Node as SaplingNode,
+        Note as SaplingNote, Nullifier as SaplingNullifier, PaymentAddress as SaplingAddress,
+        SaplingIvk,
+    },
+    transaction::{
+        components::{
+            sapling::{
+                Authorization as SaplingAuthorization, Authorized as SaplingAuthorized,
+                Bundle as SaplingBundle, GrothProofBytes,
+            },
+            Amount, OutputDescription, SpendDescription,
+        },
+        Transaction, TxId,
+    },
+    zip32::{
+        ExtendedFullViewingKey as SaplingExtendedFullViewingKey,
+        ExtendedSpendingKey as SaplingExtendedSpendingKey,
+    },
+};
+use zingoconfig::Network;
+
+pub(crate) trait ToBytes<const N: usize> {
+    fn to_bytes(&self) -> [u8; N];
+}
+
+impl ToBytes<32> for SaplingNullifier {
+    fn to_bytes(&self) -> [u8; 32] {
+        self.0
+    }
+}
+
+impl ToBytes<32> for OrchardNullifier {
+    fn to_bytes(&self) -> [u8; 32] {
+        OrchardNullifier::to_bytes(*self)
+    }
+}
+
+impl ToBytes<512> for Memo {
+    fn to_bytes(&self) -> [u8; 512] {
+        *self.encode().as_array()
+    }
+}
+
+impl ToBytes<512> for MemoBytes {
+    fn to_bytes(&self) -> [u8; 512] {
+        *self.as_array()
+    }
+}
+
+impl<const N: usize> ToBytes<N> for [u8; N] {
+    fn to_bytes(&self) -> [u8; N] {
+        *self
+    }
+}
+
+pub(crate) trait ShieldedOutputExt<P: Parameters, D: Domain>:
+    ShieldedOutput<D, ENC_CIPHERTEXT_SIZE>
+{
+    fn domain(&self, height: BlockHeight, parameters: P) -> D;
+    fn out_ciphertext(&self) -> [u8; 80];
```

Comment:
Why doesn't `zcash_note_encryption::ShieldedOutput` provide interfaces to these "encrypted transaction output" components?

Reply:
Where by "these" I mean:

* out_ciphertext
* value_commitment

Reply:
No reason I'm aware of. Note that adding required methods to a pub trait is a breaking change.

Reply:
Issued here: https://github.com/zingolabs/zingolib/issues/63